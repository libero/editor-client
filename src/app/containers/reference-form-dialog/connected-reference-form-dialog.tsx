import React, { useCallback } from 'react';
import { useDispatch } from 'react-redux';

import * as manuscriptActions from 'app/actions/manuscript.actions';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { ReferenceFormDialog } from 'app/containers/reference-form-dialog/reference-form-dialog';
import { Reference } from 'app/models/reference';

interface ConnectedReferenceFormDialogProps {
  reference?: Reference;
}

export const ConnectedReferenceFormDialog: React.FC<ConnectedReferenceFormDialogProps> = (props) => {
  const dispatch = useDispatch();
  const isNewReference = !props.reference;

  const closeDialog = useCallback(() => {
    dispatch(manuscriptEditorActions.hideModalDialog());
  }, [dispatch]);

  const handleAccept = useCallback(
    (reference: Reference) => {
      if (isNewReference) {
        dispatch(manuscriptActions.addReferenceAction(reference));
      } else {
        dispatch(manuscriptActions.updateReferenceAction(reference));
      }
      closeDialog();
    },
    [closeDialog, dispatch, isNewReference]
  );

  const handleDelete = useCallback(
    (reference: Reference) => {
      dispatch(manuscriptActions.deleteReferenceAction(reference));
      closeDialog();
    },
    [closeDialog, dispatch]
  );

  return (
    <ReferenceFormDialog
      reference={props.reference}
      onAccept={handleAccept}
      onDelete={handleDelete}
      onCancel={closeDialog}
    />
  );
};
