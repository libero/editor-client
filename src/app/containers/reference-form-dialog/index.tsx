import React, { useCallback, useState, ChangeEvent } from 'react';
import { useDispatch } from 'react-redux';

import { createBlankReference, Reference, ReferencePerson, ReferenceType } from 'app/models/reference';
import { Select } from 'app/components/select';
import { useReferenceFormStyles } from 'app/containers/reference-form-dialog/styles';
import { ActionButton } from 'app/components/action-button';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import * as manuscriptActions from 'app/actions/manuscript.actions';
import { renderConfirmDialog } from 'app/components/prompt-dialog';
import { ReferenceAuthorsList } from 'app/containers/reference-form-dialog/reference-authors-list';

interface ReferenceFormDialogProps {
  reference?: Reference;
}

export const ReferenceFormDialog: React.FC<ReferenceFormDialogProps> = ({ reference }) => {
  const classes = useReferenceFormStyles();
  const [isConfirmShown, setConfirmSnow] = useState<boolean>(false);
  const isNewReference = !reference;
  const [userReference, setReference] = useState<Reference>(reference || createBlankReference());
  const dispatch = useDispatch();

  const handleReferenceTypeChange = useCallback(
    (event: ChangeEvent<{ name: string; value: ReferenceType }>) => {
      setReference({
        ...userReference,
        type: event.target['value']
      });
    },
    [userReference]
  );

  const closeDialog = useCallback(() => {
    dispatch(manuscriptEditorActions.hideModalDialog());
  }, [dispatch]);

  const handleAuthorsListChange = useCallback(
    (refAuthors: ReferencePerson[]) => {
      setReference({
        ...userReference,
        authors: refAuthors
      });
    },
    [userReference, setReference]
  );

  const handleDone = useCallback(() => {
    userReference.authors = userReference.authors.filter(
      (author) => author['firstName'] || author['lastName'] || author['groupName']
    );
    if (isNewReference) {
      dispatch(manuscriptActions.addReferenceAction(userReference));
    } else {
      dispatch(manuscriptActions.updateReferenceAction(userReference));
    }
    closeDialog();
  }, [userReference, closeDialog, dispatch, isNewReference]);

  const handleReject = useCallback(() => {
    setConfirmSnow(false);
  }, [setConfirmSnow]);

  const handleDelete = useCallback(() => {
    setConfirmSnow(true);
  }, [setConfirmSnow]);

  const handleAccept = useCallback(() => {
    setConfirmSnow(false);
    dispatch(manuscriptActions.deleteReferenceAction(userReference));
    closeDialog();
  }, [closeDialog, dispatch, userReference]);

  return (
    <section className={classes.root}>
      <Select
        className={classes.inputField}
        name="type"
        test-id={'ref-type'}
        placeholder="Please select"
        fullWidth
        blankValue={undefined}
        label="Reference type"
        value={userReference.type}
        onChange={handleReferenceTypeChange}
        options={[
          { label: 'Journal Article', value: 'journal' },
          { label: 'Periodical Article', value: 'periodical' },
          { label: 'Book', value: 'book' },
          { label: 'Report', value: 'report' },
          { label: 'Data', value: 'data' },
          { label: 'Web Article', value: 'web' },
          { label: 'Pre-print Article', value: 'preprint' },
          { label: 'Software', value: 'software' },
          { label: 'Conference proceedings', value: 'confproc' },
          { label: 'Thesis', value: 'thesis' },
          { label: 'Patent', value: 'patent' }
        ]}
      />
      <ReferenceAuthorsList
        test-id={'ref-authors'}
        refAuthors={userReference.authors}
        onChange={handleAuthorsListChange}
      />
      <div className={classes.buttonPanel}>
        {!isNewReference ? <ActionButton variant="outlinedWarning" onClick={handleDelete} title="Delete" /> : undefined}
        <div aria-hidden={true} className={classes.spacer}></div>
        <ActionButton variant="secondaryOutlined" onClick={closeDialog} title="Cancel" />
        <ActionButton variant="primaryContained" onClick={handleDone} title="Done" />
      </div>
      {isConfirmShown
        ? renderConfirmDialog(
            'You are deleting a reference',
            'Deleting a a reference can leave unlinked citations. Are you sure you want to proceed?',
            handleAccept,
            handleReject
          )
        : undefined}
    </section>
  );
};
