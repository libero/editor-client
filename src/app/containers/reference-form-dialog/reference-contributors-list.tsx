import React, { useCallback, SyntheticEvent } from 'react';
import { TextField, IconButton, Menu, MenuItem } from '@material-ui/core';
import { has, isEqual } from 'lodash';
import MoreVertIcon from '@material-ui/icons/MoreVert';
import { SortableHandle, SortableContainer, SortableElement } from 'react-sortable-hoc';
import DragIcon from 'app/assets/drag-indicator-grey.svg';

import { ReferenceContributor } from 'app/models/reference';
import { useReferenceContributorStyles } from 'app/containers/reference-form-dialog/styles';
import { ActionButton } from 'app/components/action-button';

interface ReferenceRefContributorsListProps {
  refContributors: ReferenceContributor[];
  addCtaLabel: string;
  className: string;
  entityName: string;
  onChange(refContributors: ReferenceContributor[]): void;
}

interface RefContributorInputProps {
  value: {
    contributor: ReferenceContributor;
    entityName: string;
  };
  onDelete(): void;
  onChange(refContributors: ReferenceContributor): void;
}

const labelProps = { shrink: true };

const isGroupRefContributor = (refContributor: ReferenceContributor) => {
  return has(refContributor, 'groupName');
};

const DragHandle = React.memo(
  SortableHandle(() => <img src={DragIcon} alt="drag handle" aria-hidden={true} className="drag-handle" />),
  isEqual
);

const SortableList = React.memo(
  SortableContainer(({ className, children }) => <section className={className}> {children} </section>),
  isEqual
);

const renderRefContributorForm = (classes, refContributor, handleFormChange) => (
  <>
    <TextField
      fullWidth
      name="lastName"
      label="Surname(s)"
      classes={{ root: classes.lastName }}
      InputLabelProps={labelProps}
      variant="outlined"
      value={refContributor['lastName']}
      onChange={handleFormChange}
    />
    <TextField
      fullWidth
      name="firstName"
      label="Given name"
      InputLabelProps={labelProps}
      variant="outlined"
      value={refContributor['firstName']}
      onChange={handleFormChange}
    />
  </>
);

const renderGroupForm = (classes, refGroup, handleFormChange) => (
  <TextField
    fullWidth
    name="groupName"
    label="Group Name"
    InputLabelProps={labelProps}
    variant="outlined"
    value={refGroup['groupName']}
    onChange={handleFormChange}
  />
);

const RefContributorInput = SortableElement<RefContributorInputProps>(({ value, onChange, onDelete }) => {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
  const { contributor, entityName } = value;
  const handleFormChange = useCallback(
    (event: SyntheticEvent) => {
      onChange({ ...contributor, [event.target['name']]: event.target['value'] });
    },
    [contributor, onChange]
  );

  const openMenu = useCallback(
    (event: React.MouseEvent<HTMLButtonElement>) => {
      setAnchorEl(event.currentTarget);
    },
    [setAnchorEl]
  );

  const closeMenu = useCallback(() => {
    setAnchorEl(null);
  }, [setAnchorEl]);

  const handleToggleRefContributorType = useCallback(() => {
    if (isGroupRefContributor(contributor)) {
      onChange({ firstName: '', lastName: contributor['groupName'] });
    } else {
      onChange({ groupName: `${contributor['lastName']} ${contributor['firstName']} ` });
    }
    closeMenu();
  }, [contributor, onChange, closeMenu]);

  const handleDelete = useCallback(() => {
    onDelete();
    closeMenu();
  }, [closeMenu, onDelete]);

  const classes = useReferenceContributorStyles();

  return (
    <section className={classes.contributorInputFields}>
      <DragHandle />
      {isGroupRefContributor(contributor)
        ? renderGroupForm(classes, contributor, handleFormChange)
        : renderRefContributorForm(classes, contributor, handleFormChange)}
      <IconButton onClick={openMenu} classes={{ root: classes.menuButton }}>
        <MoreVertIcon fontSize="small" />
      </IconButton>
      <Menu anchorEl={anchorEl} keepMounted open={Boolean(anchorEl)} onClose={closeMenu}>
        <MenuItem test-id={'toggle-contributor-type-btn'} onClick={handleToggleRefContributorType}>
          {isGroupRefContributor(contributor) ? `Set to individual ${entityName}` : `Set to group ${entityName}`}
        </MenuItem>
        <MenuItem test-id={'delete-contributor-btn'} onClick={handleDelete}>
          Delete {entityName}
        </MenuItem>
      </Menu>
    </section>
  );
});

export const ReferenceContributorsList: React.FC<ReferenceRefContributorsListProps> = ({
  addCtaLabel,
  className,
  entityName,
  refContributors,
  onChange
}) => {
  const handleRefContributorChange = useCallback(
    (index: number) => (refContributor: ReferenceContributor) => {
      const updatedRefContributorsList = [...refContributors];
      updatedRefContributorsList[index] = refContributor;
      onChange(updatedRefContributorsList);
    },
    [onChange, refContributors]
  );

  const addNewRefContributor = useCallback(() => {
    onChange([...refContributors, { firstName: '', lastName: '' }]);
  }, [onChange, refContributors]);

  const handleRefContributorDelete = useCallback(
    (index: number) => () => {
      const updatedRefContributorList = [...refContributors];
      updatedRefContributorList.splice(index, 1);
      onChange(updatedRefContributorList);
    },
    [refContributors, onChange]
  );

  const onSortEnd = useCallback(
    ({ oldIndex, newIndex }) => {
      const refContributorsList = [...refContributors];

      const movedRefContributor = refContributorsList[oldIndex];
      refContributorsList.splice(oldIndex, 1);
      refContributorsList.splice(newIndex, 0, movedRefContributor);
      onChange(refContributorsList);
    },
    [onChange, refContributors]
  );

  return (
    <SortableList onSortEnd={onSortEnd} axis="y" className={className} useDragHandle={true}>
      {refContributors.map((contributor, index) => (
        <RefContributorInput
          index={index}
          key={index}
          value={{ contributor, entityName }}
          onChange={handleRefContributorChange(index)}
          onDelete={handleRefContributorDelete(index)}
        />
      ))}
      <ActionButton variant="addEntity" title={addCtaLabel} onClick={addNewRefContributor} />
    </SortableList>
  );
};
