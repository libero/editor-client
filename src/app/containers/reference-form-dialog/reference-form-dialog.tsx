import React, { useCallback, useState, ChangeEvent } from 'react';
import { get, has, pick, set } from 'lodash';
import DeleteIcon from '@material-ui/icons/Delete';
import classNames from 'classnames';
import { IconButton } from '@material-ui/core';
import { EditorState } from 'prosemirror-state';
import { Alert } from '@material-ui/lab';

import { Reference, ReferenceContributor, ReferenceInfoType } from '../../models/reference';
import { Select } from '../../components/select';
import { useReferenceFormStyles } from './styles';
import { ActionButton } from '../../components/action-button';
import { renderConfirmDialog } from '../../components/prompt-dialog';
import { ReferenceContributorsList } from './reference-contributors-list';
import { FormControlConfigType, getFormConfigForType } from './reference-forms.config';
import { renderFormControl } from './reference-form-renderer';
import refFormGrid from '../../styles/form-grid.module.scss';
import { objectsEqual } from '../../utils/view.utils';
import { ReferenceType } from '../../models/reference-type';

interface ReferenceFormDialogProps {
  reference?: Reference;
  onAccept(reference: Reference): void;
  onCancel(): void;
  onDelete(reference: Reference): void;
}

export const ReferenceFormDialog: React.FC<ReferenceFormDialogProps> = ({
  reference,
  onAccept,
  onCancel,
  onDelete
}) => {
  const classes = useReferenceFormStyles();
  const [isConfirmShown, setConfirmShow] = useState<boolean>(false);
  const isNewReference = !reference;
  const [userReference, setReference] = useState<Reference>(reference || new Reference());
  const [missingFieldsInfo, setMissingFieldsInfo] = useState<Partial<ReferenceInfoType>>({});
  const [missingFieldsConfig, setMissingFieldsConfig] = useState<Record<string, FormControlConfigType>>({});
  const handleReferenceTypeChange = useCallback(
    (event: ChangeEvent<{ name: string; value: ReferenceType }>) => {
      const newRef = userReference.clone();
      newRef.type = event.target['value'];
      newRef.referenceInfo = transferValues(userReference.referenceInfo, newRef.referenceInfo, missingFieldsInfo);
      if (userReference.referenceInfo) {
        const missingInfo = getDiffFieldValues(userReference.referenceInfo, newRef.referenceInfo, missingFieldsInfo);
        setMissingFieldsInfo(missingInfo);
        setMissingFieldsConfig(getDiffFieldsConfig(userReference.type, newRef.type, missingFieldsConfig, missingInfo));
      }
      setReference(newRef);
    },
    [userReference, missingFieldsInfo, missingFieldsConfig]
  );

  const handleAuthorsListChange = useCallback(
    (refAuthors: ReferenceContributor[]) => {
      const newRef = userReference.clone();
      newRef.authors = refAuthors;
      setReference(newRef);
    },
    [userReference, setReference]
  );

  const handleRefInfoChange = useCallback(
    (name, value) => {
      const newRef = userReference.clone();
      set(newRef.referenceInfo, name, value);
      setReference(newRef);
    },
    [userReference, setReference]
  );

  const handleMissingFieldsInfoChange = useCallback(
    (name, value) => {
      setMissingFieldsInfo({
        ...missingFieldsInfo,
        [name]: value
      });
    },
    [missingFieldsInfo]
  );

  const handleDone = useCallback(() => {
    userReference.authors = userReference.authors.filter(
      (author) => author['firstName'] || author['lastName'] || author['groupName']
    );
    if (!objectsEqual(userReference, reference)) {
      onAccept(userReference);
    } else {
      onCancel();
    }
  }, [userReference, reference, onAccept, onCancel]);

  const handleReject = useCallback(() => {
    setConfirmShow(false);
  }, [setConfirmShow]);

  const handleDeleteRow = useCallback(
    (fieldName: string) => () => {
      const updateMissingFields = { ...missingFieldsInfo };
      delete updateMissingFields[fieldName];
      setMissingFieldsInfo(updateMissingFields);
    },
    [missingFieldsInfo]
  );

  const handleDelete = useCallback(() => {
    setConfirmShow(true);
  }, [setConfirmShow]);

  const handleAccept = useCallback(() => {
    setConfirmShow(false);
    onDelete(userReference);
  }, [userReference, onDelete]);

  const formConfig = getFormConfigForType(userReference.type);
  const form = formConfig
    ? Object.entries(formConfig).map(([key, config]) => {
        return renderFormControl(
          config.type,
          config.label,
          key,
          classNames(config.className),
          get(userReference, `referenceInfo.${key}`),
          handleRefInfoChange
        );
      })
    : undefined;

  const diffForm = Object.entries(missingFieldsInfo).map(([key, value]) => {
    const config = missingFieldsConfig[key];
    return (
      <div className={classNames(classes.missingFieldsRow, classes.inputField)}>
        {renderFormControl(
          config.type,
          config.label,
          key,
          classes.missingField,
          value,
          handleMissingFieldsInfoChange,
          true
        )}
        <IconButton classes={{ root: classes.deleteButton }} onClick={handleDeleteRow(key)}>
          <DeleteIcon fontSize="small" />
        </IconButton>
      </div>
    );
  });

  return (
    <section className={classes.root}>
      <Select
        className={classes.inputField}
        name="type"
        test-id={'ref-type'}
        placeholder="Please select"
        fullWidth
        blankValue={undefined}
        label="Reference type"
        value={userReference.type}
        onChange={handleReferenceTypeChange}
        options={[
          { label: 'Journal Article', value: 'journal' },
          { label: 'Book', value: 'book' },
          { label: 'Data', value: 'data' },
          { label: 'Software', value: 'software' },
          { label: 'Preprint', value: 'preprint' },
          { label: 'Web Article', value: 'web' },
          { label: 'Conference proceedings', value: 'confproc' },
          { label: 'Report', value: 'report' },
          { label: 'Thesis', value: 'thesis' },
          { label: 'Patent', value: 'patent' },
          { label: 'Periodical Article', value: 'periodical' }
        ]}
      />
      <ReferenceContributorsList
        className={classes.inputField}
        addCtaLabel={'Add Author'}
        entityName={'author'}
        test-id={'ref-authors'}
        refContributors={userReference.authors}
        onChange={handleAuthorsListChange}
      />
      <div className={classNames(classes.inputField, refFormGrid.container)}>{form}</div>
      {diffForm.length > 0 ? (
        <div className={classes.missingFieldsSection}>
          <Alert elevation={0} variant="filled" severity="error" classes={{ root: classes.missingFieldsMessage }}>
            Unsupported elements for reference type
          </Alert>
          {diffForm}
        </div>
      ) : undefined}
      <div className={classes.buttonPanel}>
        {!isNewReference ? <ActionButton variant="outlinedWarning" onClick={handleDelete} title="Delete" /> : undefined}
        <div aria-hidden={true} className={classes.spacer}></div>
        <ActionButton variant="secondaryOutlined" onClick={onCancel} title="Cancel" />
        <ActionButton variant="primaryContained" onClick={handleDone} title="Done" />
      </div>
      {isConfirmShown
        ? renderConfirmDialog(
            'You are deleting a reference',
            'Deleting a reference will remove all linked citations from this article. Are you sure you want to proceed?',
            handleAccept,
            handleReject
          )
        : undefined}
    </section>
  );
};

function transferValues(
  prevRefInfo: ReferenceInfoType,
  nextRefInfo: ReferenceInfoType,
  prevMissingInfo: Partial<ReferenceInfoType> = {}
): ReferenceInfoType {
  Object.entries(nextRefInfo).forEach(([key, value]) => {
    nextRefInfo[key] = get(prevRefInfo, key, get(prevMissingInfo, key, value));
  });

  return nextRefInfo;
}

function getDiffFieldValues(
  prevRefInfo: ReferenceInfoType,
  nextRefInfo: ReferenceInfoType,
  prevMissingInfo: Partial<ReferenceInfoType> = {}
): Partial<ReferenceInfoType> {
  const diffKeys = Object.keys(prevRefInfo).filter((key) => !has(nextRefInfo, key) && !isValueEmpty(prevRefInfo[key]));
  const unsalvagedFieldKeys = Object.keys(prevMissingInfo).filter((key) => !has(nextRefInfo, key));
  return {
    ...pick(prevRefInfo, diffKeys),
    ...pick(prevMissingInfo, unsalvagedFieldKeys)
  };
}

function getDiffFieldsConfig(
  prevType: ReferenceType,
  nextType: ReferenceType,
  prevMissingFieldsConfig: Record<string, FormControlConfigType> = {},
  currentMissingFields: Partial<ReferenceInfoType> = {}
): Record<string, FormControlConfigType> {
  const prevConfig = getFormConfigForType(prevType);
  const nextConfig = getFormConfigForType(nextType);
  const diffKeys = Object.keys(prevConfig).filter((key: ReferenceType) => !Boolean(nextConfig[key]));
  const unsalvagedFieldsConfigKeys = Object.keys(currentMissingFields).filter((key) =>
    has(prevMissingFieldsConfig, key)
  );
  return {
    ...pick(prevConfig, diffKeys),
    ...pick(prevMissingFieldsConfig, unsalvagedFieldsConfigKeys)
  };
}

function isValueEmpty(value: EditorState | Array<ReferenceContributor> | string | number): boolean {
  if (value instanceof EditorState) {
    return !value.doc || value.doc.childCount === 0 || value.doc.textContent.trim().length === 0;
  }

  if (Array.isArray(value)) {
    return value.length === 0;
  }

  return !value;
}
