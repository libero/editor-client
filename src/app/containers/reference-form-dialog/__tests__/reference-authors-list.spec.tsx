import React from 'react';
import { create, act } from 'react-test-renderer';

import { ReferenceContributorsList } from 'app/containers/reference-form-dialog/reference-contributors-list';
import { ActionButton } from 'app/components/action-button';

jest.mock('@material-ui/core', () => ({
  Menu: ({ children }) => <div data-cmp="Menu">{children}</div>,
  Select: ({ children }) => <div data-cmp="Index">{children}</div>,
  MenuItem: ({ children }) => <div data-cmp="MenuItem">{children}</div>,
  FormControl: ({ children }) => <div data-cmp="FormControl">{children}</div>,
  InputLabel: ({ children }) => <div data-cmp="InputLabel">{children}</div>,
  TextField: () => <div data-cmp="TextField"></div>,
  IconButton: () => <div data-cmp="iconButton"></div>,
  Button: () => <div data-cmp="Button"></div>
}));

describe('Reference Authors List', () => {
  const AUTHORS = [
    { firstName: 'Test 1', lastName: 'Test 2' },
    { firstName: 'Test 3', lastName: 'Test 4' },
    { groupName: 'Test 55' }
  ];

  it('should render a list', () => {
    const wrapper = create(<ReferenceContributorsList refAuthors={AUTHORS} onChange={jest.fn()} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should add a new author entry', () => {
    const onChange = jest.fn();
    const wrapper = create(<ReferenceContributorsList refAuthors={AUTHORS} onChange={onChange} />);
    act(() => {
      wrapper.root.findByType(ActionButton).props['onClick']();
    });
    expect(wrapper).toMatchSnapshot();
    expect(onChange).toHaveBeenCalledWith([...AUTHORS, { firstName: '', lastName: '' }]);
  });

  it('should delete author', () => {
    const onChange = jest.fn();
    const wrapper = create(<ReferenceContributorsList refAuthors={AUTHORS} onChange={onChange} />);
    act(() => {
      wrapper.root.findAllByProps({ 'test-id': 'delete-author-btn' })[0].props['onClick']();
    });
    expect(wrapper).toMatchSnapshot();
    expect(onChange).toHaveBeenCalledWith([AUTHORS[1], AUTHORS[2]]);
  });

  it('should update author', () => {
    const onChange = jest.fn();
    const wrapper = create(<ReferenceContributorsList refAuthors={AUTHORS} onChange={onChange} />);
    act(() => {
      wrapper.root
        .findAllByProps({ name: 'firstName' })[0]
        .props['onChange']({ target: { name: 'firstName', value: 'ABC' } });
    });
    expect(onChange).toHaveBeenCalledWith([{ firstName: 'ABC', lastName: 'Test 2' }, AUTHORS[1], AUTHORS[2]]);
  });

  it('should toggle author type to group author', () => {
    const onChange = jest.fn();
    const wrapper = create(<ReferenceContributorsList refAuthors={AUTHORS} onChange={onChange} />);
    act(() => {
      wrapper.root.findAllByProps({ 'test-id': 'toggle-author-type-btn' })[0].props['onClick']();
    });
    expect(wrapper).toMatchSnapshot();
    expect(onChange).toHaveBeenCalledWith([{ groupName: 'Test 1 Test 2' }, AUTHORS[1], AUTHORS[2]]);
  });

  it('should toggle author type to individual author', () => {
    const onChange = jest.fn();
    const wrapper = create(<ReferenceContributorsList refAuthors={AUTHORS} onChange={onChange} />);
    act(() => {
      wrapper.root.findAllByProps({ 'test-id': 'toggle-author-type-btn' })[2].props['onClick']();
    });
    expect(wrapper).toMatchSnapshot();
    expect(onChange).toHaveBeenCalledWith([AUTHORS[0], AUTHORS[1], { firstName: '', lastName: 'Test 55' }]);
  });

  it('should rearrange authors', () => {
    const onChange = jest.fn();
    const wrapper = create(<ReferenceContributorsList refAuthors={AUTHORS} onChange={onChange} />);
    act(() => {
      const onSortEnd = wrapper.root.find((n) => n.props['onSortEnd']).props['onSortEnd'];
      onSortEnd({ oldIndex: 0, newIndex: 1 });
    });
    expect(onChange).toHaveBeenCalledWith([AUTHORS[1], AUTHORS[0], AUTHORS[2]]);
  });
});
