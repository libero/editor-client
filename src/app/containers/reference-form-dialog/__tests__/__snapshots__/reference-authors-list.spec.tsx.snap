// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reference Authors List should add a new author entry 1`] = `
<section>
   
  <section
    className="makeStyles-authorInputFields-1"
  >
    <img
      alt="drag handle"
      aria-hidden={true}
      className="drag-handle"
      src="drag-indicator-grey.svg"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="iconButton"
    />
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Set to group author
      </div>
      <div
        data-cmp="MenuItem"
      >
        Delete author
      </div>
    </div>
  </section>
  <section
    className="makeStyles-authorInputFields-1"
  >
    <img
      alt="drag handle"
      aria-hidden={true}
      className="drag-handle"
      src="drag-indicator-grey.svg"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="iconButton"
    />
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Set to group author
      </div>
      <div
        data-cmp="MenuItem"
      >
        Delete author
      </div>
    </div>
  </section>
  <section
    className="makeStyles-authorInputFields-1"
  >
    <img
      alt="drag handle"
      aria-hidden={true}
      className="drag-handle"
      src="drag-indicator-grey.svg"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="iconButton"
    />
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Set to individual author
      </div>
      <div
        data-cmp="MenuItem"
      >
        Delete author
      </div>
    </div>
  </section>
  <div
    data-cmp="Button"
  />
   
</section>
`;

exports[`Reference Authors List should delete author 1`] = `
<section>
   
  <section
    className="makeStyles-authorInputFields-1"
  >
    <img
      alt="drag handle"
      aria-hidden={true}
      className="drag-handle"
      src="drag-indicator-grey.svg"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="iconButton"
    />
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Set to group author
      </div>
      <div
        data-cmp="MenuItem"
      >
        Delete author
      </div>
    </div>
  </section>
  <section
    className="makeStyles-authorInputFields-1"
  >
    <img
      alt="drag handle"
      aria-hidden={true}
      className="drag-handle"
      src="drag-indicator-grey.svg"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="iconButton"
    />
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Set to group author
      </div>
      <div
        data-cmp="MenuItem"
      >
        Delete author
      </div>
    </div>
  </section>
  <section
    className="makeStyles-authorInputFields-1"
  >
    <img
      alt="drag handle"
      aria-hidden={true}
      className="drag-handle"
      src="drag-indicator-grey.svg"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="iconButton"
    />
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Set to individual author
      </div>
      <div
        data-cmp="MenuItem"
      >
        Delete author
      </div>
    </div>
  </section>
  <div
    data-cmp="Button"
  />
   
</section>
`;

exports[`Reference Authors List should render a list 1`] = `
<section>
   
  <section
    className="makeStyles-authorInputFields-1"
  >
    <img
      alt="drag handle"
      aria-hidden={true}
      className="drag-handle"
      src="drag-indicator-grey.svg"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="iconButton"
    />
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Set to group author
      </div>
      <div
        data-cmp="MenuItem"
      >
        Delete author
      </div>
    </div>
  </section>
  <section
    className="makeStyles-authorInputFields-1"
  >
    <img
      alt="drag handle"
      aria-hidden={true}
      className="drag-handle"
      src="drag-indicator-grey.svg"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="iconButton"
    />
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Set to group author
      </div>
      <div
        data-cmp="MenuItem"
      >
        Delete author
      </div>
    </div>
  </section>
  <section
    className="makeStyles-authorInputFields-1"
  >
    <img
      alt="drag handle"
      aria-hidden={true}
      className="drag-handle"
      src="drag-indicator-grey.svg"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="iconButton"
    />
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Set to individual author
      </div>
      <div
        data-cmp="MenuItem"
      >
        Delete author
      </div>
    </div>
  </section>
  <div
    data-cmp="Button"
  />
   
</section>
`;

exports[`Reference Authors List should toggle author type to group author 1`] = `
<section>
   
  <section
    className="makeStyles-authorInputFields-1"
  >
    <img
      alt="drag handle"
      aria-hidden={true}
      className="drag-handle"
      src="drag-indicator-grey.svg"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="iconButton"
    />
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Set to group author
      </div>
      <div
        data-cmp="MenuItem"
      >
        Delete author
      </div>
    </div>
  </section>
  <section
    className="makeStyles-authorInputFields-1"
  >
    <img
      alt="drag handle"
      aria-hidden={true}
      className="drag-handle"
      src="drag-indicator-grey.svg"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="iconButton"
    />
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Set to group author
      </div>
      <div
        data-cmp="MenuItem"
      >
        Delete author
      </div>
    </div>
  </section>
  <section
    className="makeStyles-authorInputFields-1"
  >
    <img
      alt="drag handle"
      aria-hidden={true}
      className="drag-handle"
      src="drag-indicator-grey.svg"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="iconButton"
    />
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Set to individual author
      </div>
      <div
        data-cmp="MenuItem"
      >
        Delete author
      </div>
    </div>
  </section>
  <div
    data-cmp="Button"
  />
   
</section>
`;

exports[`Reference Authors List should toggle author type to individual author 1`] = `
<section>
   
  <section
    className="makeStyles-authorInputFields-1"
  >
    <img
      alt="drag handle"
      aria-hidden={true}
      className="drag-handle"
      src="drag-indicator-grey.svg"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="iconButton"
    />
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Set to group author
      </div>
      <div
        data-cmp="MenuItem"
      >
        Delete author
      </div>
    </div>
  </section>
  <section
    className="makeStyles-authorInputFields-1"
  >
    <img
      alt="drag handle"
      aria-hidden={true}
      className="drag-handle"
      src="drag-indicator-grey.svg"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="iconButton"
    />
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Set to group author
      </div>
      <div
        data-cmp="MenuItem"
      >
        Delete author
      </div>
    </div>
  </section>
  <section
    className="makeStyles-authorInputFields-1"
  >
    <img
      alt="drag handle"
      aria-hidden={true}
      className="drag-handle"
      src="drag-indicator-grey.svg"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="iconButton"
    />
    <div
      data-cmp="Menu"
    >
      <div
        data-cmp="MenuItem"
      >
        Set to individual author
      </div>
      <div
        data-cmp="MenuItem"
      >
        Delete author
      </div>
    </div>
  </section>
  <div
    data-cmp="Button"
  />
   
</section>
`;
